python/src/spark_rapids_ml/umap.py

```diff
@@ -0,0 +1,524 @@
+    |[-0.3837125, 6.826891, -4.35618, -9.582829, -1.5456663] |
+    |[2.5012932, 4.2080708, 3.5172815, 2.5741744, -6.291008] |
+    |[9.317718, 1.3419528, -4.832837, 4.5362573, 6.9357944]  |
+    |[-6.65039, -5.438729, -6.858565, 2.9733503, 3.99863]    |
+    +--------------------------------------------------------+
+    only showing top 10 rows
+
+    >>> umap_estimator = UMAP(sample_fraction=0.5, num_workers=3).setFeaturesCol("features")
+    >>> umap_model = umap_estimator.fit(df)
+    >>> output = umap_model.transform(df).toPandas()
+    >>> embedding = cp.asarray(output["embedding"].to_list())
+    >>> print("First 10 embeddings:")
+    >>> print(embedding[:10])
+
+    First 10 embeddings:
+    [[  5.378397    6.504756 ]
+    [ 12.531521   13.946098 ]
+    [ 11.990916    6.049594 ]
+    [-14.175631    7.4849815]
+    [  7.065363  -16.75355  ]
+    [  1.8876278   1.0889664]
+    [  0.6557462  17.965862 ]
+    [-16.220764   -6.4817486]
+    [ 12.476492   13.80965  ]
+    [  6.823325  -16.71719  ]]
+
+    """
+
+    def __init__(self, sample_fraction: float = 1.0, **kwargs: Any) -> None:
+        super().__init__()
+        self.set_params(**kwargs)
+        self.sample_fraction = sample_fraction
+        self.maxRecordsPerBatch = 10000
+
+    def _create_pyspark_model(self, result: Row) -> _CumlModel:
+        raise NotImplementedError("UMAP does not support model creation from Row")
+
+    def _fit(self, dataset: DataFrame) -> "UMAPModel":
+        if self.sample_fraction < 1.0:
+            data_subset = dataset.sample(
+                withReplacement=False,
+                fraction=self.sample_fraction,
+                seed=self.cuml_params["random_state"],
+            )
+        else:
+            data_subset = dataset
+
+        input_num_workers = self.num_workers
+        # Force to single partition, single worker
+        self._num_workers = 1
+        if data_subset.rdd.getNumPartitions() != 1:
+            data_subset = data_subset.repartition(1)
+
+        maxRecordsPerBatch_str = _get_spark_session().conf.get(
+            "spark.sql.execution.arrow.maxRecordsPerBatch", "10000"
+        )
+        assert maxRecordsPerBatch_str is not None
+        self.maxRecordsPerBatch = int(maxRecordsPerBatch_str)
+
+        pipelined_rdd = self._call_cuml_fit_func(
+            dataset=data_subset,
+            partially_collect=False,
+            paramMaps=None,
+        )
+        rows = pipelined_rdd.collect()
+        # Collect and concatenate row-by-row fit results
+        from itertools import chain
+
+        embeddings = list(chain.from_iterable([row["embedding_"] for row in rows]))
+        raw_data = list(chain.from_iterable([row["raw_data_"] for row in rows]))
+        del rows
+
+        model = UMAPModel(
+            embedding_=embeddings,
+            raw_data_=raw_data,
+            n_cols=len(raw_data[0]),
+            dtype="np.float32" if isinstance(raw_data[0][0], float) else "np.float64",
```

Comment:
Looks like python float is generally double precision.   No single precision in python (def. on my linux ws, but probably vast majority of systems):  Can try this https://docs.python.org/3.10/library/sys.html#sys.float_info

Reply:
Was just playing with this in a notebook. No matter the decimal places provided, python will allocate double-precision space for the floats. As long as we're returning lists, don't think there's a way to avoid it.